jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/PluginConfigurationProcessor.java

```diff
@@ -620,13 +621,27 @@ static JavaContainerBuilder getJavaContainerBuilderWithBaseImage(
 
     if (projectProperties.getMajorJavaVersion() >= 9
         || rawConfiguration.getExpandClasspathDependencies()) {
-      List<String> dependencies =
-          projectProperties
-              .getDependencies()
+      List<Path> jars = projectProperties.getDependencies();
+      List<String> duplicates =
+          jars.stream()
+              .map(path -> path.getFileName().toString())
+              .collect(Collectors.groupingBy(filename -> filename, Collectors.counting()))
+              .entrySet()
               .stream()
-              .map(path -> appRoot.resolve("libs").resolve(path.getFileName()).toString())
+              .filter(entry -> entry.getValue() > 1)
+              .map(Map.Entry::getKey)
               .collect(Collectors.toList());
-      classpath.addAll(dependencies);
+
+      for (Path jar : jars) {
+        // Handle duplicates by appending filesize to the end of the file. This renaming logic
+        // must be in sync with the code that does the same in the other place. See
+        // https://github.com/GoogleContainerTools/jib/issues/3331
+        String jarName = jar.getFileName().toString();
+        if (duplicates.contains(jarName)) {
+          jarName = jarName.replaceFirst("\\.jar$", "-" + Files.size(jar)) + ".jar";
+        }
```

Comment:
This situation is pretty rare but what if we end up with jars that are of the same size? How about we append incremented integers to jar name? Such as `jar-name-1.jar` and `jar-name-2.jar`.

Reply:
I remember the team had a discussion on this (although might not be really deeply). There probably were multiple reasons that we converged on this method. I think one of the main reasons was that, you can identify which dependency is which by looking at the classpath string. And I think it's not just rare; I don't think anyone will ever run into the situation where two dependencies have the same artifact ID, version, and the JAR file size. I think the chance in practice is really zero.

Reply:
I see. You're right, the probability of this happening is very small.